sklearn/utils/stats.py

```diff
@@ -40,33 +45,41 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
         sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
+
+    # Sort `array` and `sample_weight` along axis=0; set `sample_weight` for nan input
+    # to 0:
     sorted_idx = np.argsort(array, axis=0)
     sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+    sorted_nan_mask = np.take_along_axis(np.isnan(array), sorted_idx, axis=0)
+    sorted_weights[sorted_nan_mask] = 0
 
-    # Find index of median prediction for each sample
+    # Compute the weighted cumulative distribution function (CDF) based on
+    # `sample_weight` and scale `percentile` along it:
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
     adjusted_percentile = percentile / 100 * weight_cdf[-1]
```

Comment:
This might lead to a division by 0 and cause a numpy warning.

We should probably protect against this:


```
weight_sum = weight_cdf[-1].copy()
weight_sum[weight_sum == 0] = 1  # Avoid division by 0.
adjusted_percentile = percentile / 100 * weight_sum
```

Reply:
I have tried to see this warning with the old code with this input:

```
arr2D = np.array([[6,5],[np.nan,1],[4,3]])
weights2D = np.array([[1,0],[1,0],[1,0]])
percentile = _weighted_percentile(arr2D, weights2D)
```
where `weight_cdf` looks like this
```
array([[1., 0.],
       [2., 0.],
       [2., 0.]])
```
and it doesn't raise a warning. `percentile / 100 * weight_cdf[-1]` would be  `array([1., 0.])`.
I think that is because we didn't include the 0 into the division with brackets like this `percentile / (100 * weight_cdf[-1])` .

`np.nanpercentile` in contrast raises: `RuntimeWarning: invalid value encountered in divide`. Maybe they did it differently there.

Should we still protect our code here with this measure?